### ISSUE TITLE
 snippets.nvim support do not work correctly

### ISSUE THREAD

**Issue by carlitux** [nvim-lua/completion-nvim#322]
As was described in this PR https://github.com/nvim-lua/completion-nvim/pull/321#issuecomment-757534341 snippets.nvim has support but it is not expanding well the snippets

**My testing minimal init.lua**

```lua
local api = vim.api
local cmd = vim.cmd
local fn = vim.fn
local luv = vim.loop
local g = vim.g

local packer_path = luv.os_homedir() .. "/.local/share/nvim/site/pack/packer/opt/packer.nvim"

if fn.filereadable(packer_path .. "/README.md") == 0 then
  cmd [[echohl WarningMsg]]
  cmd [[echom "Installing dein plugin manager, please wait"]]
  cmd([[!git clone https://github.com/wbthomason/packer.nvim ]] .. packer_path)
  cmd [[echohl None]]
end

cmd [[packadd packer.nvim]]
cmd [[set completeopt=menuone,noinsert,noselect]]

local packer = require("packer")

packer.startup(
  function()
    packer.use({"wbthomason/packer.nvim", opt = true})
    packer.use(
      {
        "norcalli/snippets.nvim",
        config = function()
          local snippets = require "snippets"
          snippets.use_suggested_mappings()
        end
      }
    )
    packer.use(
      {
        "neovim/nvim-lspconfig",
        config = function()
          local lspconfig = require "lspconfig"
          local completion = require "completion"

          local configs = require "lspconfig/configs"
          local capabilities = vim.lsp.protocol.make_client_capabilities()
          capabilities.textDocument.completion.completionItem.snippetSupport = true
          lspconfig.html.setup {on_attach = require "completion".on_attach, capabilities = capabilities}
        end,
        requires = {
          {"nvim-lua/completion-nvim"}
        }
      }
    )
  end
)

```

run `:PackerSync`
close and open neovim again and run `:checkhealth`
`sudo npm install -g vscode-html-languageserver-bin`

![Screenshot from 2021-01-10 21-09-28](https://user-images.githubusercontent.com/40538/104140552-426c8d00-5388-11eb-84ba-7a53e94eb0a4.png)

**How to reproduce**
Here see how to reproduce, snippet
![test](https://user-images.githubusercontent.com/40538/104140625-b7d85d80-5388-11eb-9239-e74c6c9975bf.gif)

**Expected behavior**
Needs to expand correctly the snippet

**Comment by runiq** [comment]
Thanks for the effort and the minimal config, that should make debugging much easier! I'll get right on it :)

**Comment by runiq** [comment]
I can reproduce with any langserver, doesn't work with sumneko and rust-analyzer either.

**Comment by runiq** [comment]
Got it I believe, dumb mistake on my part.

Could you add `vim.g.completion_enable_snippet = 'snippets.nvim'` to your minimal `init.lua` and see if that fixes things for now? @haorenW1025

**Comment by runiq** [comment]
Fix is up.

**Comment by carlitux** [comment]
@runiq this does not fix the issue for HTML at least.

`opt.get_option('enable_snippet')` and `vim.g.completion_enable_snippet` both have the same value.

Doing a little debug I found this:

Reply from HTML lsp

```
["completion_item"] = {
    ["label"] = "/html",
    ["kind"] = 10,
    ["filterText"] = "  </html>",
    ["insertTextFormat"] = 1,
    ["textEdit"] = {
        ["range"] = {
            ["end"] = {
                ["line"] = 1,
                ["character"] = 4,
            },
            ["start"] = {
                ["line"] = 1,
                ["character"] = 0,
            },
        },
        ["newText"] = "</html>",
    },
},

```

a format that is supported as I can read here https://github.com/nvim-lua/completion-nvim/blob/master/lua/completion/source/lsp.lua#L19-L26

Printing complete item:

```
["menu"] = "",
["info"] = " ",
["kind"] = "Property",
["word"] = "</html>",
["abbr"] = "/html",
["user_data"] = {
    ["lsp"] = {
        ["completion_item"] = {
            ["label"] = "/html",
            ["kind"] = 10,
            ["filterText"] = "  </html>",
            ["insertTextFormat"] = 1,
            ["textEdit"] = {
                ["range"] = {
                    ...
                },
                ["newText"] = "</html>",
            },
        },
    },
},
```

I would think the issue is here https://github.com/nvim-lua/completion-nvim/blob/master/lua/completion.lua#L129 as `completed_item.user_data.actual_item` is `nil`

**Comment by runiq** [comment]
Hmm, alright. Can you reopen this issue? (I can't) I'll take another look.

**Comment by carlitux** [comment]
I can't, too

**Comment by hrsh7th** [comment]
It is not a snippet problem.

It is TextEdit expansion problem.

**Comment by hrsh7th** [comment]
The vim-lsp has an implementation for it that is closest to VSCode does.


The vim-vsnip-integ has an implementation for it that is a simplified version of VSCode does.
https://github.com/hrsh7th/vim-vsnip-integ/blob/master/autoload/vsnip_integ.vim#L192

**Comment by runiq** [comment]
`user_data.actual_item` should only matter for snippets.nvim snippets itself, not for LSP snippets. I'll take another look later today.

**Comment by runiq** [comment]
@hrsh7th is correct, this is not a snippets problem, since `insertTextFormat` is 1 (i.e. plain text).

@carlitux I feel like I misunderstood what you expected to get originally – at first I thought you wanted something like this (`|` being the cursor):

```
html|
```

to complete to:

```
<html>
  |
</html>
```

…but judging from the `TextEdit` you posted, that's not what you're expecting to happen, right? What you really want is:

```
/html|
```

should complete to:

```
</html>|
```

Is that correct?

**Comment by hrsh7th** [comment]
No. In this case, html-language-server returns CompletionItem that represents `indentation fixing` by TextEdit.

1. Type `<`
```
<|>
```

2. Insert tag name and leave `>`
```
<html>|
```

3. Insert `<CR>` and insert closing tag
```
<html>
  </html|>
```

4. Confirm `/html` item then fix the indentation.
```
<html>
</html|>
```

The completion behavior is very complex.
Almost language servers aim to support VSCode's behavior...

**Comment by carlitux** [comment]
on the last line with text `<` after selecting completion must be `</html>` but instead is `</</html>`

final must be

```
<html>
</html>|
```

but getting

```
<html>
</</html>|
```

Not sure if has to handle indentation.

**Comment by runiq** [comment]
That's because neither `<` nor `/` are valid keyword characters for html, and completion-nvim searches for the completion keyword [by using the pattern `\k*$`](https://github.com/nvim-lua/completion-nvim/blob/74331ddc579a8b405e17532fac7ff77a8ac962ec/lua/completion/source.lua#L142-L159). In order to get the correct completion behavior, you'll have to `setlocal iskeyword+=<,/,!>

**Comment by carlitux** [comment]
@runiq maybe shouldn't be set automatically?

```lua print(vim.inspect(vim.lsp.buf_get_clients()[1].server_capabilities.completionProvider.triggerCharacters)) ```

or should be handled by `snippets.nvim` as `vim-vsnip` does with `vim-vsnip-integ`?

**Comment by runiq** [comment]
This has nothing to do with how snippets.nvim does things.

**Comment by hrsh7th** [comment]
I think snippets.nvim can do the same thing that vim-vsnip does.

But vim-vsnip-integ is handling the more complex case of LSP CompletionItem expansion.

I think it would be good to support LSP CompletionItem expansion in `completion-nvim`.
If it has done, we don't have to install vim-vsnip and vim-vsnip-integ to support complex expansion.

**Comment by hrsh7th** [comment]
I can add snippets.nvim support to vim-vsnip-integ (it is a bit weird. but will work fine may be).

**Comment by runiq** [comment]
> But vim-vsnip-integ is handling the more complex case of LSP CompletionItem expansion.
>
> I think it would be good to support LSP CompletionItem expansion in completion-nvim.

I definitely agree, I'd have expected it to be handled already. I found where that stuff was handled yesterday, I'll take a stab at it over the weekend. This should then also be able to complete multiline items.

**Comment by runiq** [comment]
> maybe shouldn't be set automatically?

@carlitux I don't think so, since this is specific to the `html` filetype and the (rather idiosyncratic IMHO) behavior of this specific langserver. So I suggest just doing `setlocal iskeyword+=<,>,!,/` in `~/.config/nvim/ftplugin/html.vim`. If you only ever use this langserver for the HTML filetype, this should solve this issue, I assume?

**Comment by eddyekofo94** [comment]
`Error executing vim.schedule lua callback: stack overflow`

This is what I am getting :/

**Comment by runiq** [comment]
@eddyekofo94 When? Where? How?

**Comment by carlitux** [comment]
> > maybe shouldn't be set automatically?
> 
> @carlitux I don't think so, since this is specific to the `html` filetype and the (rather idiosyncratic IMHO) behavior of this specific langserver. So I suggest just doing `setlocal iskeyword+=<,>,!,/` in `~/.config/nvim/ftplugin/html.vim`. If you only ever use this langserver for the HTML filetype, this should solve this issue, I assume?

not sure as you posted the video

> That's because neither `<` nor `/` are valid keyword characters for html, and completion-nvim searches for the completion keyword [by using the pattern `\k*$`](https://github.com/nvim-lua/completion-nvim/blob/74331ddc579a8b405e17532fac7ff77a8ac962ec/lua/completion/source.lua#L142-L159). In order to get the correct completion behavior, you'll have to `setlocal iskeyword+=<,/,>`:
> 
>  Peek.2021-01-14.14-31.mp4

instead of 

```html
<html>
</html>|
```

I get

```html
<html>
<//html>|
```

The same issue I am getting with flow type for example nvim state `variable.|` and suggest `?.field`


instead of 

```javascript
variable?.field|
```
I get

```html
variable.?.field|
```

What I think already it is a completion issue for these edges

**Comment by runiq** [comment]
> not sure as you posted the video

What do you mean?

> instead of
>
> ```html
> <html>
> </html>|
> ```
>
> I get
>
> ```html
> <html>
> <//html>|
> ```

Are you sure you set `iskeyword` correctly in the html document? (See video) Can you do show me `iskeyword` and the value of `v:completed_item` after that faulty completion?

> The same issue I am getting with flow type

I don't know what that is, are you referring to the [Flow typechecker](https://flow.org)? I'd need to see at least the value of `v:completed_item` after the failing completion in order to be able to debug that. Otherwise I can't even tell if this is related to snippets.nvim or textEdit expansion. (If it has `insertTextFormat = 1`, it's not a snippet, and therefore not related to snippets.nvim.)

> What I think already it is a completion issue for these edges

True, but I can't tell if it's related to snippets.nvim. Unless this has specifically to do with snippets expansion with snippets.nvim, I'd like to ask you to open a new issue.

Please don't get me wrong, I don't want to dismiss your issue, because it definitely is a valid one! I would just like to keep *this one* issue—the one I have at least a chance of solving, because I'm familiar with the code integrating snippets.nvim—on topic.

**Comment by hrsh7th** [comment]
I've created an example implementation.

https://github.com/hrsh7th/vim-vital-vs/pull/6/files
